### ISSUE TITLE
Error when checkpointing model with CyclicLR scheduler: `TypeError: cannot pickle 'WeakMethod' object`

### ISSUE THREAD

**Issue by davidgilbertson** [Lightning-AI/lightning#15901]
### Bug description

In a module that uses a CyclicLR scheduler, training fails with the error `TypeError: cannot pickle 'WeakMethod' object`.

This is not caused by an issue with Lightning, but it took me quite a while to track down the issue, so perhaps there could be a check and/or nice warning message? Maybe the system should fall back to a simpler type of checkpoint (weights only) in cases where there's an error, or some un-picklable component?

### How to reproduce the bug

```python
1. Use a `torch.optim.lr_scheduler.CyclicLR` scheduler in a `pl.LightningModule`.
2. Train as usual (with `enable_checkpointing` defaulting to `True`)
3. After a bit Lightning will attempt to save a checkpoint, which will include the scheduler, but this scheduler includes a `WeakRef` (https://github.com/pytorch/pytorch/blob/4648baa91155718642937ec27b04196fde846566/torch/optim/lr_scheduler.py#L1226, in the PyTorch code) which can't be pickled.
```

### Error messages and logs

```
Traceback (most recent call last):
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3378, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-fa864ace7871>", line 209, in <module>
    trainer.fit(
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 582, in fit
    call._call_and_handle_interrupt(
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/call.py", line 38, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 624, in _fit_impl
    self._run(model, ckpt_path=self.ckpt_path)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1061, in _run
    results = self._run_stage()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1140, in _run_stage
    self._run_train()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1163, in _run_train
    self.fit_loop.run()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py", line 200, in run
    self.on_advance_end()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py", line 295, in on_advance_end
    self.trainer._call_callback_hooks("on_train_epoch_end")
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1343, in _call_callback_hooks
    fn(self, self.lightning_module, *args, **kwargs)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 312, in on_train_epoch_end
    self._save_topk_checkpoint(trainer, monitor_candidates)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 371, in _save_topk_checkpoint
    self._save_none_monitor_checkpoint(trainer, monitor_candidates)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 660, in _save_none_monitor_checkpoint
    self._save_checkpoint(trainer, filepath)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 374, in _save_checkpoint
    trainer.save_checkpoint(filepath, self.save_weights_only)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1904, in save_checkpoint
    self._checkpoint_connector.save_checkpoint(filepath, weights_only=weights_only, storage_options=storage_options)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/checkpoint_connector.py", line 540, in save_checkpoint
    self.trainer.strategy.save_checkpoint(_checkpoint, filepath, storage_options=storage_options)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/strategies/strategy.py", line 466, in save_checkpoint
    self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/lightning_lite/plugins/io/torch_io.py", line 54, in save_checkpoint
    _atomic_save(checkpoint, path)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/lightning_lite/utilities/cloud_io.py", line 67, in _atomic_save
    torch.save(checkpoint, bytesbuffer)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/torch/serialization.py", line 423, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/torch/serialization.py", line 635, in _save
    pickler.dump(obj)
TypeError: cannot pickle 'WeakMethod' object
```

### Environment

Version: `1.8.3.post1`

### More info

Explicitly passing `ModelCheckpoint(save_weights_only=True),` to the `Trainer.callbacks` parameter avoids the error.

cc @awaelchli

**Comment by Borda** [comment]
@davidgilbertson so you say (to reproduce it) that just switching the scheduler to `CyclicLR` with the Boring model would yield this error?

**Comment by davidgilbertson** [comment]
Yep, sorry, I didn't do a repro.

Here 'tis with just `configure_optimizers` changed:

```py
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.CyclicLR(
            optimizer=optimizer,
            base_lr=0.01,
            max_lr=0.1,
        )
        return [optimizer], [lr_scheduler]
        


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()
```

**Comment by awaelchli** [comment]
I have observed this issue with torch>=1.13. You can downgrade to 1.12 to avoid it for now.

**Comment by khadiravana-belagavi** [comment]
the issue still persists in lightning == 2.0.1

**Comment by awaelchli** [comment]
@khadiravana-belagavi This was never an issue with Lightning. See here: https://github.com/Lightning-AI/lightning/issues/15901#issuecomment-1341918688
To get the fix, you need to upgrade torch to >= 2.0. From the linked commit in PyTorch, you can see that it was included in that release: https://github.com/pytorch/pytorch/commit/f5e20d60604481cb6f8625d2fce92e7e76c6dffc (it is tagged with [v2.0.0](https://github.com/pytorch/pytorch/releases/tag/v2.0.0))

**Comment by dhruvbird** [comment]
The issue still exists for me. Getting `cannot pickle 'WeakMethod' object` when I start using CyclicLR scheduler.
